### PR TITLE
MEJO-16 Communicate with all the RGB sensors at the same time

### PR DESCRIPTION
**Pull request by U30RK4**

### CODE REVIEW

File: Nucleo_F746ZG_app/Src/test.c [hunk 1]
```diff
@@ -71,15 +71,16 @@ void TestTask(void const * argument)
...
-    memset(rgb_number, '\0', sizeof(char) * 16);
+    char rgb_number[25];
+    memset(rgb_number, '\0', sizeof(char) * 25);
 
-    for (int i = 0; i < 15; ++i)
+    for (int i = 0; i < 8;++i)
     {
-      sprintf(rgb_number, "R%d G%d B%d C%d \r\n", rgb_colors(RED),rgb_colors(GREEN), rgb_colors(BLUE), rgb_colors(CLEAR));
+      sprintf(rgb_number,"R:%d G:%d B:%d C:%d \r\n",rgb_colors(RED),rgb_colors(GREEN), rgb_colors(BLUE), rgb_colors(CLEAR));
       HAL_UART_Transmit(&huart3, (uint8_t*) rgb_number, strlen(rgb_number),0xFFFF);
```
Comment (lengyeltamas) [inline Nucleo_F746ZG_app/Src/test.c]: async_log! :)

File: Nucleo_F746ZG_app/Src/test.c [hunk 2]
```diff
@@ -71,15 +71,16 @@ void TestTask(void const * argument)
...
-    char rgb_number[16];
-    memset(rgb_number, '\0', sizeof(char) * 16);
+    char rgb_number[25];
+    memset(rgb_number, '\0', sizeof(char) * 25);
 
-    for (int i = 0; i < 15; ++i)
+    for (int i = 0; i < 8;++i)
     {
-      sprintf(rgb_number, "R%d G%d B%d C%d \r\n", rgb_colors(RED),rgb_colors(GREEN), rgb_colors(BLUE), rgb_colors(CLEAR));
+      sprintf(rgb_number,"R:%d G:%d B:%d C:%d \r\n",rgb_colors(RED),rgb_colors(GREEN), rgb_colors(BLUE), rgb_colors(CLEAR));
```
Comment (lengyeltamas) [inline Nucleo_F746ZG_app/Src/test.c]: a tesztben most több szenzort is tesztelsz egymás után?

File: Nucleo_F746ZG_app/Src/rgb_sensor.c [hunk 3]
```diff
@@ -1,10 +1,15 @@
 #include "stm32f7xx.h"
 #include "rgb_sensor.h"
 #include "i2c.h"
+#include "usart.h"
+#include "string.h"
+#include <stdio.h>
+
+volatile rgb_colors_t current_colors[8];
```
Comment (johnyka) [inline Nucleo_F746ZG_app/Src/rgb_sensor.c]: does it have to be volatile?

File: Nucleo_F746ZG_app/Src/rgb_sensor.c [hunk 4]
```diff
@@ -23,7 +28,7 @@ void select_mux_channel(uint8_t channel)
 {
```
Comment (johnyka) [inline Nucleo_F746ZG_app/Src/rgb_sensor.c]: you don't use this function if I'm not mistaken

File: Nucleo_F746ZG_app/Src/rgb_sensor.c [hunk 5]
```diff
@@ -81,16 +85,84 @@ uint16_t rgb_colors(rgb_color_t color)
...
+  case CHANNEL5:
+    (multiplexer_channel <<= 5);
+    break;
+  case CHANNEL6:
+    (multiplexer_channel <<= 6);
+    break;
+  case CHANNEL7:
+    (multiplexer_channel <<= 7);
+    break;
+  case ALLCHANNELS:
```
Comment (johnyka) [inline Nucleo_F746ZG_app/Src/rgb_sensor.c]: for me it doesn't really make sense. You just iterate over the channels, and enable all of them but do nothing with the sensor. This is completely pointless thing to do

File: Nucleo_F746ZG_app/Src/rgb_sensor.c [hunk 6]
```diff
@@ -81,16 +85,84 @@ uint16_t rgb_colors(rgb_color_t color)
...
+  HAL_GPIO_WritePin(GPIOG, SB_nEXP_RESET_Pin, SET);
+  uint8_t multiplexer_channel = 0b00000001;
+
+  switch (channel) {
+  default:
+  case CHANNEL0:
+    multiplexer_channel = 0b00000001;
+    break;
+  case CHANNEL1:
+    (multiplexer_channel <<= 1);
```
Comment (johnyka) [inline Nucleo_F746ZG_app/Src/rgb_sensor.c]: You can simplify this switch case statement simply just shifting the bits with the channel parameter casted to an integer

File: Nucleo_F746ZG_app/Src/rgb_sensor.c [hunk 7]
```diff
@@ -81,16 +85,84 @@ uint16_t rgb_colors(rgb_color_t color)
...
+    current_colors[(int) channel].clear_color = rgb_colors(CLEAR);
+
+  } else if (channel == ALLCHANNELS) {
+
+    for (int i = 0; i < 8; ++i) {
+      current_colors[i].red_color = rgb_colors(RED);
+      current_colors[i].green_color = rgb_colors(GREEN);
+      current_colors[i].blue_color = rgb_colors(BLUE);
+      current_colors[i].clear_color = rgb_colors(CLEAR);
+      (multiplexer_channel <<= 1);
```
Comment (johnyka) [inline Nucleo_F746ZG_app/Src/rgb_sensor.c]: I don't think this function works. You shift the multiplexer channel but don;t actually switch to that channel, you only modify the variable. In order to switch channel you have to transmit this to the multiplexer address. I think what happens here is that you query the color 8 times from the same sensor.

File: Nucleo_F746ZG_app/Src/rgb_sensor.c [hunk 8]
```diff
@@ -81,16 +79,61 @@ uint16_t rgb_colors(rgb_color_t color)
...
+
+  uint8_t color_high = color_temp;
+  rgb_color = ((uint16_t) color_high << 8) | color_low;
 
   return rgb_color;
 }
+
+void select_multiplexer_channel(int channel)
+{
+  HAL_GPIO_WritePin(GPIOG, SB_nEXP_RESET_Pin, SET);
```
Comment (johnyka) [inline Nucleo_F746ZG_app/Src/rgb_sensor.c]: This has to be done only once in rgb_init

File: Nucleo_F746ZG_app/Src/rgb_sensor.c [hunk 9]
```diff
@@ -81,16 +79,61 @@ uint16_t rgb_colors(rgb_color_t color)
...
+    for (int i = 0; i < 8; ++i) {
+
+      select_multiplexer_channel(i);
+
+      current_colors[i].red_color = rgb_colors(RED);
+      current_colors[i].green_color = rgb_colors(GREEN);
+      current_colors[i].blue_color = rgb_colors(BLUE);
+      current_colors[i].clear_color = rgb_colors(CLEAR);
+
+      async_log("R:%d G:%d B:%d C:%d\r\n", current_colors[i].red_color,
```
Comment (johnyka) [inline Nucleo_F746ZG_app/Src/rgb_sensor.c]: remove this log

File: Nucleo_F746ZG_app/Src/test.c [hunk 10]
```diff
@@ -35,7 +36,7 @@ void TestTask(void const * argument)
   {
     // semi-automatic test implementation BEGIN here
 
-    BSP_LED_Toggle(LED_GREEN);
+    /*BSP_LED_Toggle(LED_GREEN);
```
Comment (johnyka) [inline Nucleo_F746ZG_app/Src/test.c]: this should not be commented out

File: Nucleo_F746ZG_app/Src/test.c [hunk 11]
```diff
@@ -70,14 +71,11 @@ void TestTask(void const * argument)
...
-    memset(rgb_number, '\0', sizeof(char) * 16);
+    osDelay(1000);*/
 
-    for (int i = 0; i < 15; ++i)
-    {
-      sprintf(rgb_number, "R%d G%d B%d C%d \r\n", rgb_colors(RED),rgb_colors(GREEN), rgb_colors(BLUE), rgb_colors(CLEAR));
-      HAL_UART_Transmit(&huart3, (uint8_t*) rgb_number, strlen(rgb_number),0xFFFF);
+    for (int i = 0; i < 8; ++i) {
+      select_multiplexer_channel(i);
+      async_log("%d: R:%d G:%d B:%d C:%d\r\n",i,rgb_colors(RED),rgb_colors(GREEN),rgb_colors(BLUE),rgb_colors(CLEAR));
```
Comment (johnyka) [inline Nucleo_F746ZG_app/Src/test.c]: you should print from your rgb_colors structure and not calling your functions here

**Comment by U30RK4** [comment]
merge conflict solved

File: Nucleo_F746ZG_app/Src/test.c [hunk 12]
```diff
@@ -73,14 +73,11 @@ void TestTask(void const * argument)
...
-    char rgb_number[16];
-    memset(rgb_number, '\0', sizeof(char) * 16);
 
-    for (int i = 0; i < 15; ++i)
-    {
-      sprintf(rgb_number, "R%d G%d B%d C%d \r\n", rgb_colors(RED),rgb_colors(GREEN), rgb_colors(BLUE), rgb_colors(CLEAR));
-      HAL_UART_Transmit(&huart3, (uint8_t*) rgb_number, strlen(rgb_number),0xFFFF);
-      osDelay(500);
+    for (int i = 0; i < 8; ++i) {
+      select_multiplexer_channel(i);
```
Comment (johnyka) [inline Nucleo_F746ZG_app/Src/test.c]: you don't have to select the multiplexer channel here, since your actual_color_storage function does it

Comment (U30RK4) [reply Nucleo_F746ZG_app/Src/test.c]: mission completed

**Pull request merged**